src/ShowingTransitionSupport.js

```diff
@@ -0,0 +1,212 @@
+* @public
+*/
+module.exports = {
+
+	/**
+	* @private
+	*/
+	name: 'ShowingTransitionSupport',
+
+	/**
+	* A read-only property for checking whether we are in the middle of a transition.
+	*
+	* @type {Boolean}
+	* @default false
+	* @public
+	*/
+	showingTransitioning: false,
+
+	/**
+	* The amount of time the "showing" transition takes to complete in milliseconds.
+	*
+	* @type {Number}
+	* @default undefined
+	* @public
+	*/
+	showingDuration: undefined,
+
+	/**
+	* The amount of time the "hiding" transition takes to complete in milliseconds.
+	*
+	* @type {Number}
+	* @default undefined
+	* @public
+	*/
+	hidingDuration: undefined,
+
+	/**
+	* The method to fire at the end of the "showing" transition. This may be name of a method on the
+	* component, or a function.
+	*
+	* @type {String|Function}
+	* @default undefined
+	* @public
+	*/
+	shownMethod: undefined,
+
+	/**
+	* The method to fire at the end of the "hiding" transition. This may be name of a method on the
+	* component, or a function.
+	*
+	* @type {String|Function}
+	* @default undefined
+	* @public
+	*/
+	hiddenMethod: undefined,
+
+	/**
+	* The the classname to apply for the "shown" (component is visible) resting state.
+	*
+	* @type {String}
+	* @default undefined
+	* @public
+	*/
+	shownClass: undefined,
+
+	/**
+	* The the classname to apply for the "hidden" (component is not visible) resting state.
+	*
+	* @type {String}
+	* @default 'hidden'
+	* @public
+	*/
+	hiddenClass: undefined,
+
+	/**
+	* The the classname to apply for the "hiding" (component has started the transition to the
+	* hidden state) transition state.
+	*
+	* @type {String}
+	* @default 'hiding'
+	* @public
+	*/
+	hidingClass: undefined,
+
+	/**
+	* The the classname to apply for the "showing" (component has started the transition to the
+	* shown state) transition state.
+	*
+	* @type {String}
+	* @default 'showing'
+	* @public
+	*/
+	showingClass: undefined,
+
+	/**
+	* Initializes the defaults, and prepares the component with its classes in case of initially
+	* "showing: false".
+	*
+	* @method
+	* @private
+	*/
+	create: kind.inherit(function (sup) {
+		return function () {
+			sup.apply(this, arguments);
+			this.showingDuration = (this.showingDuration === undefined) ? null      : this.showingDuration;
+			this.hidingDuration  = (this.hidingDuration  === undefined) ? null      : this.hidingDuration;
+			this.shownMethod     = (this.shownMethod     === undefined) ? null      : this.shownMethod;
+			this.hiddenMethod    = (this.hiddenMethod    === undefined) ? null      : this.hiddenMethod;
+			this.shownClass      = (this.shownClass      === undefined) ? null      : this.shownClass;
+			this.hiddenClass     = (this.hiddenClass     === undefined) ? 'hidden'  : this.hiddenClass;
+			this.hidingClass     = (this.hidingClass     === undefined) ? 'hiding'  : this.hidingClass;
+			this.showingClass    = (this.showingClass    === undefined) ? 'showing' : this.showingClass;
+			this.showingChanged();
```

Comment:
Ah, I see that you're deferring the super call until the callback completes so you have to override showingChanged. I'm hesitant to recommend we take this with the assumption that all impacted showing logic is within an upstream `showingChanged` method. If we've made assumptions anywhere that the change handlers for `showing` were synchronous, we'll have issues here.

I think the idea is solid and the implementation looks good otherwise. We could either mark it as WIP noting the possible issues or hold it until after 2.7 when we can look at `showing` support more broadly. I'm leaning toward the latter since this is a new feature without any consumers.

Thoughts @Djspaceg, @graynorton, @aarontam, @clinuz?